File: csharp/generate/Generate.cs
// Copyright (c) 2019 Yodlee, Inc. All Rights Reserved.
// Licensed under the MIT License. See `LICENSE` for details.

using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Security;
using System;
using System.IO;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.CommandLine;
using System.CommandLine.Invocation;
using System.CommandLine.Builder;
using System.Threading.Tasks;

namespace jwt_csharp_netcore
{
    class Generate
    {
        /*
         * 
         * generate.
         *
         * Usage:
         * generate.exe --key=<str> --issuer=<str> [--username=<str>]
         * generate.exe (-h | --help)
         *
         * Options:
         *  -k, --key=<str>          Path to private key file
         *  -i, --issuer-id=<str>    Issuer id from api connsole
         *  -u, --username=<str>     Username if generating user token         
         *  -h, --help               Show this screen.
         *
         */

        static async Task Main(string[] args)
        {
            Command command = prepareOptions(args);

            command.Handler = CommandHandler.Create<FileInfo, string, string>(CreateToken);

            await command.InvokeAsync(args);
        }

        private static void CreateToken(FileInfo key, string issuerId, string username)
        {
            long currentTime = DateTimeOffset.Now.ToUnixTimeSeconds();

            var payload = new Dictionary<string, object>();

            payload["iss"] = issuerId;
            payload["iat"] = currentTime;
            payload["exp"] = currentTime + 1800;

            if (username != null)
            {
                payload["sub"] = username;
            }

            string token = CreateToken(payload, key);

            Console.WriteLine(token);
        }

        public static string CreateToken(Dictionary<string, object> payload, FileInfo privateKey)
        {
            RSAParameters rsaParams;

            using (var tr = privateKey.OpenText())
            {
                var pemReader = new PemReader(tr);

                RsaPrivateCrtKeyParameters privkey = null;
                Object obj = pemReader.ReadObject();

                if (obj != null)
                {
                    privkey = (RsaPrivateCrtKeyParameters)obj;
                }

                rsaParams = DotNetUtilities.ToRSAParameters(privkey);
            }

            using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider())
            {
                rsa.ImportParameters(rsaParams);

                return Jose.JWT.Encode(payload, rsa, Jose.JwsAlgorithm.RS512);
            }
        }

        public static Command prepareOptions(params string[] args)
        {
            Argument parserHook = new Argument { IsHidden = true };

            parserHook.AddValidator(command => {
                if (!command.Children.Contains("--key"))
                {
                    return "Option '--key' is required.";
                }

                if (!command.Children.Contains("--issuer-id"))
                {
                    return "Option '--issuer-id' is required.";
                }

                return null; // no error
            });

            RootCommand rootCommand = new RootCommand
            {
                new Option(new[] { "--key", "-k" }, "Path to your private key file.")
                {
                    Argument = new Argument<FileInfo>().ExistingOnly()
                },
                new Option(new[] { "--issuer-id", "-i" }, "Your issuer id.")
                {
                    Argument = new Argument<string>()
                },
                new Option(new[] { "--username", "-u" }, "A user id.")
                {
                    Argument = new Argument<string>()
                },
                parserHook
            };

            rootCommand.Description = "An application to generate JWTs";

            rootCommand.TreatUnmatchedTokensAsErrors = true;

            return rootCommand;
        }
    }
}

File: node/generate/generate.js
// Copyright (c) 2019 Yodlee, Inc. All Rights Reserved.
// Licensed under the MIT License. See `LICENSE` for details.

/**
 * Usage:
 *
 * $ node generate.js --key=/path/to/your/private.key --issuer-id=issuer-id --username=username
 */

const fs = require('fs')
const jwt = require('jsonwebtoken')
const argv = require('yargs')
  .option('key', {
    alias: 'k',
    describe: 'Path to your private key'
  })
  .option('issuer-id', {
    alias: 'i',
    describe: 'Your issuer id'
  })
  .option('username', {
    alias: 'u',
    describe: 'Create a user token'
  })
  .demandOption(['key', 'issuer-id'], 'Please provide both a key file and issuer id')
  .requiresArg(['key', 'issuer-id', 'username'])
  .help()
  .argv

function generate() {
    const keyFile = argv.key

    //Location of the file with your private key
    const privateKey = fs.readFileSync(keyFile, "utf8");

    const currentTime =  Math.floor(Date.now() / 1000);

    const signOptions = {
	    algorithm: "RS512"  //Yodlee requires RS512 algorithm when encoding JWT
    };

    let payload = {};

    // The issuer id from the API Dashboard
    payload.iss = argv['issuer-id']
    payload.iat = currentTime;  //Epoch time when token is issued in seconds
    payload.exp = currentTime + 1800;  //Epoch time when token is set to expire. Must be 1800 seconds

    //generate user token if present in argv
    if (typeof argv.username != 'undefined') {
        payload.sub = argv.username
    }

    return token = jwt.sign(payload, privateKey, signOptions);
}

module.exports = generate

if (require.main === module) {
    console.log(generate())
}
